src/wallet/wallet.cpp

```diff
@@ -82,15 +82,56 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
-// Custom deleter for shared_ptr<CWallet>.
-static void ReleaseWallet(CWallet* wallet)
+static Mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cv;
+static std::set<CWallet*> g_unloading_wallet_set;
+
+// Unregister and delete the wallet right after BlockUntilSyncedToCurrentChain
+// so that it's in sync with the current chainstate.
+static void DeleteWallet(CWallet* wallet)
 {
     wallet->WalletLogPrintf("Releasing wallet\n");
     wallet->BlockUntilSyncedToCurrentChain();
     wallet->Flush();
+    UnregisterValidationInterface(wallet);
     delete wallet;
 }
 
+// Custom deleter for shared_ptr<CWallet>.
+// If the release is due to a UnloadWallet call then deletion is delegated to
```

Comment:
This comment needs to be updated with the latest change.

Also nit: s/a UnloadWallet/an UnloadWallet/